vars/unitTestPost.groovy

```diff
@@ -26,90 +26,104 @@
-  }
+    if (config['valgrind_stash']) {
+        String valgrind_pattern = config.get('valgrind_pattern',
+                                             '*memcheck.xml')
+        stash name: config['valgrind_stash'], includes: valgrind_pattern
+    }
 
-  if (stage_info['NLT']) {
-    def cb_result = currentBuild.result
-    discoverGitReferenceBuild referenceJob: config.get('referenceJobName',
-                                              'daos-stack/daos/master'),
-                              scm: 'daos-stack/daos'
-    recordIssues enabledForFailure: true,
-                 failOnError: !ignore_failure,
-                 ignoreFailedBuilds: true,
-                 ignoreQualityGate: true,
-                 // Set qualitygate to 1 new "NORMAL" priority message
-                 // Supporting messages to help identify causes of
-                 // problems are set to "LOW".
-                 qualityGates: [
-                   [threshold: 1, type: 'TOTAL_ERROR'],
-                   [threshold: 1, type: 'TOTAL_HIGH'],
-                   [threshold: 1, type: 'NEW_NORMAL', unstable: true],
-                   [threshold: 1, type: 'NEW_LOW', unstable: true]],
-                  name: "Node local testing",
-                  tool: issues(pattern: 'vm_test/nlt-errors.json',
-                               name: 'NLT results',
-                               id: 'VM_test')
+    if (stage_info['NLT']) {
+        String cb_result = currentBuild.result
+        discoverGitReferenceBuild(
+          referenceJob: config.get('referenceJobName',
+                                   'daos-stack/daos/master'),
+          scm: 'daos-stack/daos')
+        recordIssues enabledForFailure: true,
+                     failOnError: !ignore_failure,
+                     ignoreFailedBuilds: true,
+                     ignoreQualityGate: true,
+                     // Set qualitygate to 1 new "NORMAL" priority message
+                     // Supporting messages to help identify causes of
+                     // problems are set to "LOW".
+                     qualityGates: [
+                       [threshold: 1, type: 'TOTAL_ERROR'],
+                       [threshold: 1, type: 'TOTAL_HIGH'],
+                       [threshold: 1, type: 'NEW_NORMAL', unstable: true],
+                       [threshold: 1, type: 'NEW_LOW', unstable: true]],
+                      name: 'Node local testing',
+                      tool: issues(pattern: 'vm_test/nlt-errors.json',
+                                   name: 'NLT results',
+                                   id: 'VM_test')
 
-    if (cb_result != currentBuild.result) {
-      println "The recordIssues step changed result to ${currentBuild.result}."
+        if (cb_result != currentBuild.result) {
+            println(
+              "The recordIssues step changed result to ${currentBuild.result}.")
+        }
+    }
+    if (valgrind_found) {
+        error "Valgrind Errors detected in ${env.STAGE_NAME}"
```

Comment:
That's fair enough (and I've requested confirmation on that in the ticket as this conflicts directly with the original requirements gathering for our CI).  But we should not be employing `error()`, and confusingly reporting the stage as `FAILED` (i.e. the process of running the test failed), because what it really is is `UNSTABLE` (tests running was successful but some tests had failures) to achieve this.

The subsequent stages already support skipping if the current build result is `UNSTABLE`:
https://github.com/daos-stack/pipeline-lib/blob/master/vars/skipStage.groovy#L175-L188 called from https://github.com/daos-stack/pipeline-lib/blob/master/vars/skipStage.groovy#L415 and https://github.com/daos-stack/pipeline-lib/blob/master/vars/skipStage.groovy#L529.  So you just need to make the stage that has valgrind errors `UNSTABLE` to achieve the goal of stopping the pipeline for valgrind errors.

The above mechanism also allows the behaviour of stopping on `UNSTABLE` stages to be overridden by a commit pragma, so is ultimately more flexible.  It will give us an answer for developers that complain that they don't want valgrind errors to be the end of their pipeline.  It could be that gatekeeping makes this a hard requirement (due to the escape that led to SRE-1729), but we should address that in `skipStage()`.

Reply:
Changed in current push to set the stage to unstable, so a commit pragma can be used as mentioned above.